pippy/PipelineStage.py

```diff
@@ -618,16 +637,167 @@ def forward_one_chunk(
+        fwd_cache[chunk] = (
+            output_tuple,  # stage_output
+            flatten_input_tensors,  # input_values
+        )
+
+        return output, send_reqs
+
+    def forward_one_chunk_one_ipipe(
+        self,
+        chunk: int,
+        inner_rank: int,
+        args_split,
+        kwargs_split,
+        composite_args,
+        composite_kwargs,
+        fwd_cache: Dict[int, Any],
+    ):
+        if self.rank == self.nstages - 1:
+            # Need improvement- how do we properly pass targets to forward_maybe_with_nosync?
+            targets = args_split[chunk][0]
+            # print(f"[Rank{self.rank}] args_split {args_split}")
+        else:
+            targets = None
+
+        try:
+            if inner_rank == 0:
+                output = self.forward_maybe_with_nosync(
+                    inner_rank, targets, *composite_args, **composite_kwargs
+                )
+                self.pipe_cache[chunk][inner_rank] = output
+            elif (
+                inner_rank == self.inner_depth - 1
+                and self.rank == self.nstages - 1
+            ):  # last stage, last node
+                output = self.forward_maybe_with_nosync(
+                    inner_rank,
+                    targets,
+                    self.pipe_cache[chunk][self.inner_depth - 2],
+                    targets,
+                    **composite_kwargs,
+                )  # self.inner_pipe >= 2 is asserted
+                self.pipe_cache[chunk][self.inner_depth - 1] = output
+            else:
+                output = self.forward_maybe_with_nosync(
+                    inner_rank,
+                    targets,
+                    self.pipe_cache[chunk][inner_rank - 1],
+                    **composite_kwargs,
+                )
+                self.pipe_cache[chunk][inner_rank] = output
+
+        except Exception as e:
+            exc_msg = f"""
+            Rank {self.rank} failed to run forward stage: {self.name}
+            args: {map_debug_info(composite_args)}
+            kwargs: {map_debug_info(composite_kwargs)}
+            """
+            raise RuntimeError(exc_msg) from e
+
+        if inner_rank == self.inner_depth - 1:
+            # Unify output form to tuple for easy correspondance with
+            # `act_send_info`
+            output_tuple = output if type(output) is tuple else (output,)
+            send_reqs = self._send_activations(output_tuple)
+
+            # Save activations and inputs for backward
+            flat_args = flatten_args(composite_args)
+            flat_kwargs = flatten_args(composite_kwargs)
+            flatten_input_tensors = flat_args + flat_kwargs
+            fwd_cache[chunk] = (
+                output_tuple,  # stage_output
+                flatten_input_tensors,  # input_values
             )
 
-            print(f"[Rank{self.rank}] Arrived here")
+            return output, send_reqs
+        else:
+            return None, None
+
+    def forward_one_chunk_all_ipipe(
```

Comment:
nit: maybe we can think of a better name for this? for example, using `_inner_pipe`? because ipipe is so confusing to me.